### DISCUSSION TITLE
Retrieve sort overriding, along with table configuration

### DISCUSSION THREAD

**Discussion by ghost** [fraupflaume/rpivotTable#2] [Ideas]
Congratulations on the great job you have done [here](https://stackoverflow.com/a/72120956/170792)!

I would like to see just a small improvement. 
It's often useful to sort attributes (like month names) in a different way than the default (alphabetical). 
This custom sorting is not stored (or retrieved) at the moment. 

[Example](https://pivottable.js.org/examples/montreal_2014.html)

The R code looks like this:

```
...,
sorters = " 
  function(attr) {
  var sortAs = $.pivotUtilities.sortAs;
  if (attr == 'MY ATTRIBUTE') { return sortAs(['Beta', 'Gamma', 'Delta', 'Alpha']); }
}",
...
```
Thank you

**Comment by ghost** [comment]
I have only found out that the `where` object, in `giveItBack()` contains a `sorters` item with the function definition.

**Reply by fraupflaume** [reply]
I guess you answered your own question or suggestion. From the vignette, this is how you might use the parameter `sorters`:

```
sorters = "function(attr) {
                 var sortAs = $.pivotUtilities.sortAs;
                 if (attr == \"Hair\") { return sortAs([\"Red\", \"Brown\", \"Blond\", \"Black\"]); }
               }"
```

**Reply by ghost** [reply]
[gulp] 😰
This is what I have done. 
I can get the definition (as a string), by giving:

`JSON.parse(where).x.params.sorters[0];`

but I am not sure if I have used the best way to redefine sorters in the restored table. 
My last (seems to be successful) attempt was this ([reference](https://stackoverflow.com/a/32070428/170792)):

```
if(where.includes('"sorters":[')){
  gimmeMore['sorters'] = new Function('return ' + JSON.parse(where).x.params.sorters[0].replaceAll('\n',''))();
}
```

I guess I should put it inside an if-clause too (tables without sorters didn't work).  
No JavaScript experience, you see. :)

In your expert opinion, what would you suggest?

Thank you.

**Reply by fraupflaume** [reply]
I have a question—completely off-topic. If someone (a person with a significant amount of education in computer information systems) explained what a superuser was (you know, someone with admin rights) and then cited you a reference written in 1956. You would:

1. laugh in their face; they must be joking (they weren't, though)
2. cry because having heard the information, you felt that your IQ points were sucked from your head; you were literally less intelligent having listened
3. ask them to repeat it; you must have misheard them (you didn't, though)

**Reply by ghost** [reply]
4. remain silent because you think they are probably still mad at you. You have realized that your insufficient knowledge of English (along with your bad sense of humor) often leads to misunderstandings. You just wish you could find a way to apologize. :)

**Reply by fraupflaume** [reply]
Or... if you're really smart, you just pretend you don't speak English. That way, you can save your IQ points, and you no longer have to talk to this person of questionable intelligence. 
Sometimes I wish I could pretend that. (I used to work with someone who did that. I thought it was hilarious....but I wasn't allowed to laugh if they did it to someone in front of me.) 
My dad, on the other hand, pretended he only spoke English. Until you said something wrong in a language he knew, then he would tell you you said it wrong (in English). He would then go on to say, you don't know how to say it correctly, so just speak English.

**Comment by fraupflaume** [comment]
Sheesh! I should have realized this was needed! Sorry. So to figure this out, I needed to understand how it was used and how it would have been done if it was strictly the JS version of things. At first, I made it far more complicated than was necessary, smacked my forehead, then found a solution that almost bit me...sigh. That led me to the following changes to the `giveitBack()` function:

```
function giveItBack() {                           // function to regurgitate
    el = document.querySelectorAll('.rpivotTable');
    console.log("working on the giver");
    ods = [...el[0].ownerDocument.scripts];         // make it an array
    path = window.location.pathname.split("/").pop().split(".").slice()[0]; //filename
    for(j=0; j < el.length; j++){
      elId = el[j].getAttribute("id");
      where = ods.filter(function(ods){             // filter scripts for data
        return ods.dataset['for'] === elId;
      })[0].innerHTML; 
      where2 = JSON.parse(where).x.data;            // format data for pivotUI()
      where3 = HTMLWidgets.dataframeToD3(where2);   // ...still formatting
      if(window.localStorage.getItem(path + '_table' + j) === null) { // alert
    // basic system alert
        //jj = j + 1;                                                 
        //alert("WARNING: There is no saved pivot table configuration for " + path + "'s table " + jj + ".");
    // custom alert
        msg = "<b>WARNING</b><br><br>There is no saved pivot table configuration for<br>" + path + "."
        notice(msg);
    // Either Alert: from here on needed whether basic or custom alert
        continue;
      }
      gimme = window.localStorage.getItem(path + '_table' + j); // get storage
      gimmeMore = JSON.parse(gimme);                            // prepare for recall
      if(where.includes('"subtotals":true')){       // is the option 'subtotal' used?
        gimmeMore.renderers = $.pivotUtilities.subtotal_renderers;
        gimmeMore.dataClass = $.pivotUtilities.SubtotalPivotData;
      }; 
      if(where.includes('"tsv":true')){             // is the option 'tsv' used?
        gimmeMore.renderers = $.extend(gimmeMore.renderers, $.pivotUtilities.export_renderers);
      };
      if(where.includes('sortAs')){
        // passed as a function, they will get lost in save & retrieve
        stringy = $('#' + elId).data("pivotUIOptions").sorters;
        gimmeMore.sorters = stringy;
      }
      $('#' + elId).pivotUI(where3, gimmeMore, true, "en"); // put it back!
    }
  };
```

No changes to the other areas of the cookie-ish JS.

**Reply by ghost** [reply]
Another question: Do you think it could be possible, somehow, to have all the renderers available? 
Like this :

![image](https://user-images.githubusercontent.com/3167917/167820548-bd23cb80-03fc-49b1-a730-c494828bdd14.png)

P.S. I could make it a SO question. - Just name your bounty :)

**Reply by fraupflaume** [reply]
Is this a trick question? Isn't that the list of what's available now? Maybe I'm slow...I have been reading mind-numbing-try-to-stay-awake-and-focused-despite-the-torture all morning.

Nevermind. I went back and looked. I got you. The table, table-barchart, and heatmaps are replaced when subtotals are added. Let me take a look at this and see what I can come up with. (Anything to stop the torture...that has a deadline. Sigh.)

**Reply by ghost** [reply]
There's no need to hurry. This is not urgent at all.

**Reply by fraupflaume** [reply]
I spent a good bit of time going through how pivottable.js and subtotal.js come together. I'm not sure if it would be easier to rebuild subtotal.js or continue going through it. 
* It looks like there is coding to specifically delete the non-subtotal renderers. However, a lot is going on there, so I'm not 100% on that. 
* If I end up rewriting subtotal.js, would you rather I used the React.js version of pivottable.js? 
  * It's a bit prettier; it doesn't currently work with subtotal.js. 
  * I think it has fewer renderer options (it's primarily built using Plotly versus C3 or D3). I think it has a bit more functionality, though. The interactivity with Plotly is supposed to be far beyond what we get with the graphs in this version of pivottable.js. (it looks like the original author of pivottable.js works for Plotly.)
  * I _think_ that the way you implement the React vs. non-React is the same. So the user wouldn't see a difference in implementation. 

Currently, subtotal.js only supports C3 and D3. I assume that's why the R version only has those renderer options. Although, I'm a bit confused about why it would matter (it does matter, though). It's like subtotal.js rewrote pivottable.js instead of being more of an add-on. The person who wrote the add-on is _obviously_ wicked intelligent! So I'm sure there was a reason.

However I go about it, it won't be a quick turnaround.

**Reply by ghost** [reply]
At first I just want to say that I admire that you can work so much, so fast and of this quality!. I guess my motto suits you better: Я твой слуга Я твой работник.

Well. to begin with, I do not ask you to go about it. I am more than ok with the state of things right now.
Also I have zero experience on React.js. I first read about it in one of your answers, a while ago.
Personaly, I don't use the widget for graphs (nor do my end users, I guess). I like to create my own. Plotly would be better, for sure.
I find the subtotals functionality usefull.   
Ιf you decide to take this on, I am sure you'll make the best choises. (_Create a new discussion in that case_)

Oh, I also have left a comment for you [here](https://github.com/fraupflaume/rpivotTable/discussions/1#discussioncomment-2739539). What are your thoughts on it?

**Reply by ghost** [reply]
I forgot to emphasise, once more, that I would give higher priority in replacing the CRAN version of the package. 
Some time ago, while I was in desperation, I had sent a mail to CRAN (I did not know who else to turn to, since the maintainers didn't respond). Τheir answer was:

> ...please discuss with the maintainer of the package, or submit a
> package which offers an improvement.

Can you imagine how frustrated the users who download the package from CRAN feel? It is a defamation for the credibility of the R language as a whole.

**Reply by fraupflaume** [reply]
Без труда не вытащишь и рыбку из пруда LOL I studied Russian in college...but I think I've forgotten it all.
As far as submitting the package, I feel conflicted. It is undoubtedly built on the work of others. I would rather have the acknowledgment of the original developer before I took that step.

I'm going to follow that link now.

**Comment by ghost** [comment]
@fraupflaume: In order to retrieve the table with the predefined locale setting, I made this change. Do you agree?

![image](https://user-images.githubusercontent.com/107258922/174544346-c9e86ef2-cfca-44ca-8c16-736951069b92.png)

**Comment by fraupflaume** [comment]
I'm sorry for the delay! Your method works perfectly...and gives me ideas of how I might have coded this better.
